### ISSUE TITLE
SNMP V1 Trap Showing Varbind of SNMP-COMMUNITY-MIB::snmpTrapCommunity.0 which is not present in actual PDU

### ISSUE THREAD

**Issue by sbhattach** [etingof/pysnmp#78]
Hi
In PySNMP v1 TRAP receiver shows SNMP-COMMUNITY-MIB::snmpTrapCommunity.0  but it does not exist in actual V1 trap PDU.
There is no translation of SNMP v2 TRAP to V1. (rfc3418

)
Its an actual v1 trap PDU which don't have snmpTrapCommunity.0 in its var bind.( RFC 1157)

Refer:
the attached debug_log.txt
[debug_log.txt](https://github.com/etingof/pysnmp/files/1303071/debug_log.txt)
Refer v1 PDU capture.
[v1_trpa_pdu.pcapng.txt](https://github.com/etingof/pysnmp/files/1303104/v1_trpa_pdu.pcapng.txt)

used code is something like this:

```
def cbFun(snmpEngine, stateReference, contextEngineId, contextName,
          varBinds, cbCtx):
    global mibViewController
    print "####################### NEW Notification() #######################"
    execContext = snmpEngine.observer.getExecutionContext(
        'rfc3412.receiveMessage:request'
    )
    print('#Notification from %s \n#ContextEngineId: "%s" \n#ContextName: "%s" \n#SNMPVER "%s" \n#SecurityName "%s"' % ('@'.join([str(x) for x in execContext['transportAddress']]),contextEngineId.prettyPrint(),contextName.prettyPrint(), execContext['securityModel'], execContext['securityName']))
    for oid, val in varBinds:
        output = rfc1902.ObjectType(rfc1902.ObjectIdentity(oid),val).resolveWithMib(mibViewController).prettyPrint()
        print output
```
used Pysnmp version : 4.3.8

**Comment by etingof** [comment]
Thank you for rising this!

Yes, that's the intended behavior -- with pysnmp, user-facing side is always SMIv2. That is why even SNMPv1 TRAPs get translated into SNMPv2 TRAPs. Or when you are sending v1 TRAPs you still need to pass v2 TRAPs to pysnmp.

Why is it a problem? I thought of it as a feature because you get TRAPs uniformly, regardless of SNMP version your peer is talking, no?

You can convert it back to SNMPv1 if you want to:

```
from pysnmp.proto.proxy import rfc2576

v1_pdu = rfc2576.v2ToV1(v2_pdu)

```
Where `v2_pdu` is what you get from low-level pysnmp hooks. Or you could reconstruct it from the var-binds you get in the callback:

```
from pysnmp.proto.proxy import rfc2576
from pysnmp.proto.api import v2c

v2_pdu = v2c.SNMPv2TrapPDU()

v2c.TrapPDUAPI.setVarBinds(v2c_pdu, varBinds)
v1_pdu = rfc2576.v2ToV1(v2_pdu)

```

Does it make sense?

**Comment by sbhattach** [comment]
Ya it's it helpful. My concern was it should not add the community string since the original PDU don't have it, and creates confusion to end user. Now as you point out the v1 conversation I will try that out so I can avoid community string. Please close the case.

**Comment by sbhattach** [comment]
One quick question by Pysnm can I forward incoming trap/inform to other pysnmp trap/inform receiver? 
Something like Bellow,

(Agent)------(pysnmp fwd)-------(pysnmp receiver)

**Comment by etingof** [comment]
> My concern was it should not add the community string since the original PDU don't have it,
> and creates confusion to end user.

That's how it should work:

```
snmpTrapCommunity OBJECT-TYPE
    SYNTAX      OCTET STRING
    MAX-ACCESS  accessible-for-notify
    STATUS      current
    DESCRIPTION
        "The value of the community string field of an SNMPv1
         message containing a Trap PDU which is forwarded by a
         a proxy forwarder application using an SNMP version
         other than SNMPv1.  The value of this object SHOULD
         contain the value of the community string field from
         the original SNMPv1 message containing a Trap PDU as
         generated by an SNMPv1 agent.  There is no SIZE
         constraint specified for this object because RFC 1157
         does not impose any explicit limitation on the length
         of community strings (their size is constrained
         indirectly by the SNMP message size)."
    ::= { snmpCommunityMIBObjects 4 }
```

Apparently, it does not fully implemented with pysnmp -- current implementation [it does not](https://github.com/etingof/pysnmp/blob/master/pysnmp/proto/proxy/rfc2576.py#L128) carry over the community string from v1 TRAP message into v2 TRAP PDU, it just adds an empty string.

So what you have spotted looks like a bug! Let me fix that.

**Comment by etingof** [comment]
> One quick question by Pysnmp can I forward incoming trap/inform to other pysnmp 
> trap/inform receiver?

Sure, why not? You just need to send your TRAP PDU elsewhere from your TRAP receiving callback function. Here is [an example script](http://pysnmp.sourceforge.net/examples/v3arch/asyncore/proxy/command/protocol-conversion.html#snmpv2c-to-snmpv1-conversion) for proxying SNMP commands. I can add similar for TRAPs if you need more guidance with this, do you?
 
BTW, there is already the [SNMP proxy forwarder](https://snmpfwd.sourceforge.io/configuration/examples/trap-forwarding-snmpv2.html) app that can perform TRAP forwarding.

**Comment by etingof** [comment]
> Apparently, it does not fully implemented with pysnmp -- current implementation it does not carry over the community string from v1 TRAP message into v2 TRAP PDU, it just adds an empty string.

Fixed in 9d35779f1cf8dce05bb8abe12aa7a2bb0732967a
